### ISSUE TITLE
[Bug] tf.sqrt Inconsistent behaviour

### ISSUE THREAD

**Issue by Ghost---Shadow** [tensorflow/tensorflow#42602]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.5
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: V10.1.243
- GPU model and memory: RTX2070 MaxQ, 8GB

**Describe the current behavior**

When I run this code, I get the result and gradients

```py
x = tf.Variable([[1.0, 0.0]])

with tf.GradientTape() as tape:
    x = tf.square(x)
    x = tf.sqrt(x)
    z = tf.reduce_sum(x, axis=-1)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss,x))
'''
Output:
[1] 0.5
[[1 1]]
'''
```

However, when I refactor the three lines inside a function, the gradient becomes NaN

```py
@tf.function # Happens in both eager and graph
def fun(v):
    v = tf.square(v)
    v = tf.sqrt(v)
    return tf.reduce_sum(v, axis=-1)

x = tf.Variable([[1.0, 0.0]])
with tf.GradientTape() as tape:
    z = fun(x)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss, x))

'''
Output:
[1] 0.5
[[1 nan]]
'''
```

Changing `x = tf.sqrt(x)` to `x = tf.sqrt(x + 1e-7)` makes the gradient non NaN.

**Describe the expected behavior**

Either result is acceptable as long as the behaviour is consistent.

**Comment by Saduf2019** [comment]
I am able to replicate the issue faced, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/b8f0c0f7a15aec8140a6a8c3925bb309/untitled385.ipynb)

**Comment by saxenasaurabh** [comment]
The failure in `tf.function` looks reasonable but it is unclear to me how this works without `tf.function`. The problem basically is that we are unable to fold the computation `1/x * x = 1` in the gradients path. This fails when x is `0`.

**Comment by allenlavoie** [comment]
The eager version of the code overwrites the `x` Python local variable. I assume that's the only reason they're giving different results.

So far we've decided not to do anything smart with backprop regarding inverse functions canceling etc. (it's been discussed several times).

**Comment by saxenasaurabh** [comment]
While using an epsilon is a reasonable workaround here, you may also want to check out `tf.custom_gradient` to implement numerically stable gradient functions for cases such as this one.

**Comment by mdanatg** [comment]
I think the refactoring inadvertently changes the logic.

If we make sure the codes are equivalent by just removing the tf.function decorator, then the outputs seem to be actually consistent:

```
# Same result with or without @tf.function
def fun(v):
    v = tf.square(v)
    v = tf.sqrt(v)
    return tf.reduce_sum(v, axis=-1)

x = tf.Variable([[1.0, 0.0]])
with tf.GradientTape() as tape:
    z = fun(x)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss, x))
```

```
[1] 0.5
[[1 -nan]]
```

**Comment by Ghost---Shadow** [comment]
@allenlavoie It happens in both eager and graph. You can comment out the `tf.function`. I dont think this is being caused by XLA optimization.

**Comment by mdanatg** [comment]
@Ghost---Shadow the two pieces of code that you originally included don't calculate the same thing. If you correct the eager version, you will also get nan, as expected:

```
x_var = tf.Variable([[1.0, 0.0]])
x = x_var  # avoid calling tape.gradient wrt intermediate outputs

with tf.GradientTape() as tape:
    x = tf.square(x)
    x = tf.sqrt(x)
    z = tf.reduce_sum(x, axis=-1)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss, x_var))
```

**Comment by Ghost---Shadow** [comment]
@mdanatg This doesnt return NaN

```
x = tf.Variable([[1.0, 0.0]])

with tf.GradientTape() as tape:
    x = tf.square(x)
    x = tf.sqrt(x)
    z = tf.reduce_sum(x, axis=-1)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss,x))
'''
Output:
[1] 0.5
[[1 1]]
'''
```

**Comment by mdanatg** [comment]
It doesn't, and it's not expected to, because it's not taking a gradient with respect to the non-differentiable `sqrt` function. Let's rename the variables to make that clearer:

```
x = tf.Variable([[1.0, 0.0]])

with tf.GradientTape() as tape:
    x = tf.square(x)
    wrong_x = tf.sqrt(x)
    z = tf.reduce_sum(wrong_x, axis=-1)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss, wrong_x))
```

So you might as well write this with the same effect:

```
with tf.GradientTape() as tape:
    wrong_x = tf.constant([[1.0, 0.0]])
    tape.watch(wrong_x)
    z = tf.reduce_sum(wrong_x, axis=-1)
    loss = tf.nn.l2_loss(z)

tf.print(z, loss)
tf.print(tape.gradient(loss, wrong_x))
```
